Packet++/src/Asn1Codec.cpp

```diff
@@ -0,0 +1,569 @@
+		std::vector<uint8_t> result;
+		result.reserve(m_ValueLength);
+
+		for (auto record : m_SubRecords)
+		{
+			auto encodedRecord = record->encode();
+			result.insert(result.end(), std::make_move_iterator(encodedRecord.begin()), std::make_move_iterator(encodedRecord.end()));
+		}
+		return result;
+	}
+
+	Asn1SequenceRecord::Asn1SequenceRecord(const std::vector<Asn1Record*>& subRecords)
+		: Asn1ConstructedRecord(Asn1TagClass::Universal, static_cast<uint8_t>(Asn1UniversalTagType::Sequence), subRecords)
+	{}
+
+	Asn1SetRecord::Asn1SetRecord(const std::vector<Asn1Record*>& subRecords)
+			: Asn1ConstructedRecord(Asn1TagClass::Universal, static_cast<uint8_t>(Asn1UniversalTagType::Set), subRecords)
+	{}
+
+	Asn1PrimitiveRecord::Asn1PrimitiveRecord(uint8_t tagType) : Asn1Record()
+	{
+		m_TagType = tagType;
+		m_TagClass = Asn1TagClass::Universal;
+		m_IsConstructed = false;
+	}
+
+	Asn1IntegerRecord::Asn1IntegerRecord(uint32_t value) : Asn1PrimitiveRecord(static_cast<uint8_t>(Asn1UniversalTagType::Integer))
+	{
+		m_Value = value;
+
+		if (m_Value <= std::pow(2, sizeof(uint8_t) * 8))
+		{
+			m_ValueLength = sizeof(uint8_t);
+		}
+		else if (value <= std::pow(2, sizeof(uint16_t) * 8))
+		{
+			m_ValueLength = sizeof(uint16_t);
+		}
+		else if (value <= std::pow(2, 3 * 8))
+		{
+			m_ValueLength = 3;
+		}
+		else
+		{
+			m_ValueLength = sizeof(uint32_t);
+		}
+
+		m_TotalLength = m_ValueLength + 2;
+	}
+
+	void Asn1IntegerRecord::decodeValue(uint8_t* data, bool lazy)
+	{
+		switch (m_ValueLength)
+		{
+			case 1:
+			{
+				m_Value = *(uint8_t*)data;
+				break;
+			}
+			case 2:
+			{
+				m_Value = be16toh(*(uint16_t*)data);
+				break;
+			}
+			case 3:
+			{
+				uint8_t tempData[4] = {0};
+				memcpy(tempData + 1, data, 3);
+				m_Value = be32toh(*(uint32_t*)tempData);
+				break;
+			}
+			case 4:
+			{
+				m_Value = be32toh(*(uint32_t*)data);
+				break;
+			}
+			default:
+			{
+				throw std::runtime_error("An integer ASN.1 record of more than 4 bytes is not supported");
+			}
+		}
+	}
+
+	std::vector<uint8_t> Asn1IntegerRecord::encodeValue() const
+	{
+		std::vector<uint8_t> result;
+		result.reserve(m_ValueLength);
+
+		switch (m_ValueLength)
+		{
+			case 1:
+			{
+				result.push_back(static_cast<uint8_t>(m_Value));
+				break;
+			}
+			case 2:
+			{
+				uint8_t tempArr[sizeof(uint16_t)];
+				auto hostValue = htobe16(static_cast<uint16_t>(m_Value));
+				memcpy(tempArr, &hostValue, m_ValueLength);
+				std::copy(tempArr, tempArr + m_ValueLength, std::back_inserter(result));
+				break;
+			}
+			case 3:
+			{
+				uint8_t tempArr[sizeof(uint32_t)];
+				auto hostValue = htobe32(static_cast<uint32_t>(m_Value));
+				memcpy(tempArr, &hostValue, sizeof(uint32_t));
+				std::copy(tempArr + 1, tempArr + m_ValueLength + 1, std::back_inserter(result));
+				break;
+			}
+			case 4:
+			{
+				uint8_t tempArr[sizeof(uint32_t)];
+				auto hostValue = htobe32(static_cast<uint32_t>(m_Value));
+				memcpy(tempArr, &hostValue, m_ValueLength);
```

Comment:
maybe we can have consistency, since you use `memcpy(tempArr, &hostValue, sizeof(uint32_t));` for case 3?

Reply:
I'm not sure I understand your suggestion...

Reply:
Oh got it. Fixed in https://github.com/seladb/PcapPlusPlus/pull/1353/commits/155788fa5b55d836931a8df30cedc602d77cd807